File: backend/schemas/issue.py
"""
Issue 模块 Pydantic Schema 定义
AutoCraft 问题跟踪系统 - Pydantic v2 数据验证模型
"""
from datetime import datetime
from enum import Enum
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict


# ============================================================================
# 枚举定义
# ============================================================================

class IssueStatus(str, Enum):
    """问题状态枚举"""
    NEW = "新建"
    IN_PROGRESS = "处理中"
    RESOLVED = "已解决"
    CLOSED = "已关闭"


class IssuePriority(str, Enum):
    """优先级枚举"""
    HIGH = "高"
    MEDIUM = "中"
    LOW = "低"


class IssueCategory(str, Enum):
    """问题分类枚举"""
    BUG = "Bug"
    OPTIMIZATION = "优化"
    REQUIREMENT = "需求"
    OTHER = "其他"
    
    # FlowTicket 异常分类
    PYTEST_LIMIT = "pytest_limit"
    GUARDIAN_LIMIT = "guardian_limit"
    VERIFY_LIMIT = "verify_limit"
    TIMEOUT = "timeout"
    EXECUTION_TIMEOUT = "execution_timeout"
    RETRY_LIMIT = "retry_limit"
    CONFIG_MISSING = "config_missing"
    CODE_ERROR = "code_error"
    ENV_ERROR = "env_error"
    MANUAL_INTERVENTION = "manual_intervention"


# ============================================================================
# 基础模型
# ============================================================================

class IssueBase(BaseModel):
    """Issue 基础模型"""
    title: str = Field(..., max_length=100, description="问题标题")
    description: Optional[str] = Field(None, description="问题描述")
    category: IssueCategory = Field(..., description="问题分类")
    priority: IssuePriority = Field(..., description="优先级")
    related_task_id: Optional[str] = Field(None, description="关联任务单号")


# ============================================================================
# 请求模型
# ============================================================================

class IssueCreate(IssueBase):
    """创建 Issue 请求模型"""
    project_id: int = Field(..., description="项目 ID")
    stage_id: Optional[int] = Field(None, description="阶段 ID")
    workflow_id: Optional[int] = Field(None, description="工作流 ID")
    created_by: str = Field(..., description="创建人")


class IssueUpdate(BaseModel):
    """更新 Issue 请求模型（所有字段可选）"""
    title: Optional[str] = Field(None, max_length=100, description="问题标题")
    description: Optional[str] = Field(None, description="问题描述")
    category: Optional[IssueCategory] = Field(None, description="问题分类")
    priority: Optional[IssuePriority] = Field(None, description="优先级")
    status: Optional[IssueStatus] = Field(None, description="状态")
    related_task_id: Optional[str] = Field(None, description="关联任务单号")
    assignee: Optional[str] = Field(None, max_length=100, description="处理人")
    due_date: Optional[datetime] = Field(None, description="截止日期")
    assign_note: Optional[str] = Field(None, description="分配说明")


class IssueResolveRequest(BaseModel):
    """解决 Issue 请求模型"""
    solution: str = Field(..., description="解决方案")
    solution_process: Optional[str] = Field(None, description="解决过程")
    resolved_by: str = Field(..., description="解决人")


class IssueCloseRequest(BaseModel):
    """关闭 Issue 请求模型"""
    closed_by: str = Field(..., description="关闭人")
    closed_reason: Optional[str] = Field(None, description="关闭原因")


class FlowTicketIssueCreate(BaseModel):
    """FlowTicket 自动创建问题单请求模型"""
    source: str = Field(default="flow_ticket", description="来源，固定为 flow_ticket")
    error_type: str = Field(..., description="错误类型 (pytest_limit/timeout等)")
    error_message: str = Field(..., description="异常消息")
    task_id: Optional[str] = Field(None, description="关联任务ID")
    node_id: Optional[str] = Field(None, description="当前节点ID")
    recovery_info: Optional[str] = Field(None, description="恢复信息")
    project_id: int = Field(..., description="项目 ID")
    created_by: str = Field(default="system_auto", description="创建人，默认为 system_auto")


class IssueAssignRequest(BaseModel):
    """分配 Issue 请求模型"""
    assignee: str = Field(..., max_length=100, description="处理人")
    due_date: Optional[datetime] = Field(None, description="截止日期")
    assign_note: Optional[str] = Field(None, description="分配说明")


# ============================================================================
# 响应模型
# ============================================================================

class IssueResponse(IssueBase):
    """Issue 响应模型"""
    id: int = Field(..., description="Issue ID")
    issue_no: str = Field(..., description="问题编号")
    status: IssueStatus = Field(..., description="状态")
    project_id: int = Field(..., description="项目 ID")
    stage_id: Optional[int] = Field(None, description="阶段 ID")
    workflow_id: Optional[int] = Field(None, description="工作流 ID")
    assignee: Optional[str] = Field(None, description="处理人")
    assigned_at: Optional[datetime] = Field(None, description="分配时间")
    due_date: Optional[datetime] = Field(None, description="截止日期")
    assign_note: Optional[str] = Field(None, description="分配说明")
    solution_process: Optional[str] = Field(None, description="解决过程")
    solution: Optional[str] = Field(None, description="解决方案")
    resolved_by: Optional[str] = Field(None, description="解决人")
    resolved_at: Optional[datetime] = Field(None, description="解决时间")
    closed_by: Optional[str] = Field(None, description="关闭人")
    closed_at: Optional[datetime] = Field(None, description="关闭时间")
    created_by: str = Field(..., description="创建人")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")

    model_config = ConfigDict(from_attributes=True)


class IssueListResponse(BaseModel):
    """Issue 列表响应模型"""
    total: int = Field(..., description="总记录数")
    items: List[IssueResponse] = Field(..., description="Issue 列表")
    page: int = Field(..., description="当前页码")
    page_size: int = Field(..., description="每页数量")
    total_pages: int = Field(..., description="总页数")


class IssueDetailResponse(BaseModel):
    """Issue 详情响应模型"""
    data: IssueResponse = Field(..., description="Issue 详情")


# ============================================================================
# 查询参数模型
# ============================================================================

class IssueFilter(BaseModel):
    """Issue 筛选条件模型"""
    page: int = Field(default=1, ge=1, le=1000, description="页码")
    page_size: int = Field(default=10, ge=1, le=100, description="每页数量")
    status: Optional[IssueStatus] = Field(None, description="状态筛选")
    priority: Optional[IssuePriority] = Field(None, description="优先级筛选")
    category: Optional[IssueCategory] = Field(None, description="分类筛选")
    creator: Optional[str] = Field(None, description="创建人筛选")
    assignee: Optional[str] = Field(None, description="处理人筛选")
    keyword: Optional[str] = Field(None, description="关键字搜索（标题/描述）")
    project_id: Optional[int] = Field(None, description="项目 ID 筛选")
    stage_id: Optional[int] = Field(None, description="阶段 ID 筛选")
    workflow_id: Optional[int] = Field(None, description="工作流 ID 筛选")
    related_task_id: Optional[str] = Field(None, description="关联任务单号筛选")
    created_at_from: Optional[datetime] = Field(None, description="创建时间起始")
    created_at_to: Optional[datetime] = Field(None, description="创建时间结束")


# ============================================================================
# 通用响应模型
# ============================================================================

class MessageResponse(BaseModel):
    """通用消息响应模型"""
    message: str = Field(..., description="消息内容")
    issue_no: Optional[str] = Field(None, description="问题编号")

File: src/types/issue.ts
/**
 * 问题记录单类型定义
 * 用于问题跟踪和管理功能
 */

// ============================================================================
// 基础类型定义
// ============================================================================

/**
 * 问题状态枚举值
 */
export type IssueStatus = '新建' | '处理中' | '已解决' | '已关闭' | 'open' | 'resolved' | 'closed';

/**
 * 优先级枚举值
 */
export type IssuePriority = '高' | '中' | '低';

/**
 * 问题分类枚举值
 */
export type IssueCategory = 'Bug' | '优化' | '需求' | '其他';

// ============================================================================
// Element Plus Tag 类型映射
// ============================================================================

/**
 * Element Plus Tag 组件的类型
 * 用于不同状态/优先级的视觉区分
 */
export type TagType = 'success' | 'warning' | 'danger' | 'info' | 'primary' | '';

// ============================================================================
// 选项常量接口
// ============================================================================

/**
 * 下拉选项接口定义
 * 用于表单中的选择器组件
 */
export interface OptionItem {
  /** 显示标签 */
  label: string;
  /** 实际值 */
  value: string;
  /** Element Plus Tag 类型 */
  type: TagType;
}

// ============================================================================
// 核心数据模型
// ============================================================================

/**
 * 问题记录单完整接口
 * 包含 16 个字段，覆盖问题全生命周期
 */
export interface Issue {
  /** 问题单号，格式：ISSUE-YYYYMMDD-NNNN */
  id: string;
  
  /** 问题标题 */
  title: string;
  
  /** 问题描述 */
  description: string;
  
  /** 状态：新建 | 处理中 | 已解决 | 已关闭 */
  status: IssueStatus;
  
  /** 优先级：高 | 中 | 低 */
  priority: IssuePriority;
  
  /** 问题分类 */
  category: IssueCategory | '';
  
  /** 创建人 */
  creator: string;
  
  /** 创建时间，ISO 8601 格式 */
  created_at: string;
  
  /** 解决时间，ISO 8601 格式 */
  resolved_at: string;
  
  /** 解决人 */
  resolved_by: string;
  
  /** 解决方案总结 */
  solution: string;
  
  /** 根因分析 */
  root_cause: string;
  
  /** 解决过程 */
  solution_process: string;
  
  /** 经验教训 */
  lesson_learned: string;
  
  /** 关闭原因 */
  closed_reason: string;
  
  /** 关联任务单号 */
  related_task_id: string;
  
  // ========== 关联信息 ==========
  
  /** 关联项目名称 */
  project_name?: string;
  
  /** 关联阶段名称 */
  stage_name?: string;
  
  /** 关联工作流名称 */
  workflow_name?: string;
  
  // ========== 处理信息 ==========
  
  /** 处理人 */
  assignee?: string;
  
  /** 分配时间 */
  assigned_at?: string;
  
  /** 预期完成时间 */
  due_date?: string;
  
  /** 分配说明 */
  assign_note?: string;
  
  // ========== 解决信息（兼容字段） ==========
  
  /** 解决人（兼容字段） */
  resolver?: string;
  
  /** 解决方案总结（兼容字段） */
  resolution_summary?: string;
}

// ============================================================================
// 请求接口定义
// ============================================================================

/**
 * 创建问题请求接口
 * 创建时不需要填写解决相关字段
 */
export interface CreateIssueRequest {
  /** 问题标题 */
  title: string;
  
  /** 问题描述 */
  description: string;
  
  /** 优先级 */
  priority: IssuePriority;
  
  /** 问题分类 */
  category: IssueCategory | '';
  
  /** 关联任务单号（可选） */
  related_task_id?: string;
}

/**
 * 更新问题请求接口
 * 所有字段可选，支持部分更新
 */
export interface UpdateIssueRequest {
  /** 问题标题 */
  title?: string;
  
  /** 问题描述 */
  description?: string;
  
  /** 优先级 */
  priority?: IssuePriority;
  
  /** 问题分类 */
  category?: IssueCategory | '';
  
  /** 关联任务单号 */
  related_task_id?: string;
}

/**
 * 解决问题请求接口
 * 用于将问题状态从 open 更新为 resolved
 */
export interface ResolveIssueRequest {
  /** 解决人 */
  resolved_by: string;
  
  /** 解决方案总结 */
  solution: string;
  
  /** 根因分析 */
  root_cause: string;
  
  /** 解决过程 */
  solution_process: string;
  
  /** 经验教训 */
  lesson_learned: string;
  
  // ========== 兼容字段 ==========
  
  /** 解决人（兼容字段） */
  resolver?: string;
  
  /** 解决方案总结（兼容字段） */
  resolution_summary?: string;
}

/**
 * 关闭问题请求接口
 * 用于将问题状态从 resolved 更新为 closed
 */
export interface CloseIssueRequest {
  /** 关闭原因 */
  closed_reason: string;
}

// ============================================================================
// 查询与响应接口
// ============================================================================

/**
 * 问题查询参数接口
 * 用于列表分页和筛选
 */
export interface IssueQueryParams {
  /** 页码，从 1 开始 */
  page?: number;
  
  /** 每页数量 */
  page_size?: number;
  
  /** 状态筛选 */
  status?: IssueStatus;
  
  /** 优先级筛选 */
  priority?: IssuePriority;
  
  /** 分类筛选 */
  category?: IssueCategory;
  
  /** 创建人筛选 */
  creator?: string;
  
  /** 关键字搜索（标题/描述） */
  keyword?: string;
  
  /** 关联任务单号筛选 */
  related_task_id?: string;
  
  /** 创建时间起始 */
  created_at_from?: string;
  
  /** 创建时间结束 */
  created_at_to?: string;
}

/**
 * 问题列表响应接口
 */
export interface IssueListResponse {
  /** 问题列表 */
  items: Issue[];
  
  /** 总数量 */
  total: number;
  
  /** 当前页码 */
  page: number;
  
  /** 每页数量 */
  page_size: number;
  
  /** 总页数 */
  total_pages: number;
}

// ============================================================================
// 问题总结相关接口
// ============================================================================

/**
 * 问题总结接口
 */
export interface IssueSummary {
  /** 问题单号 */
  issue_no: string;
  
  /** 根因分析 */
  root_cause: string;
  
  /** 解决方案 */
  solution_approach: string;
  
  /** 经验教训 */
  lessons_learned: string;
  
  /** 预防措施 */
  prevention_measures: string;
  
  /** 创建时间 */
  created_at: string;
  
  /** 更新时间 */
  updated_at: string;
}

/**
 * 提交总结请求接口
 */
export interface SubmitSummaryRequest {
  /** 根因分析 */
  root_cause: string;
  
  /** 解决方案 */
  solution_approach: string;
  
  /** 经验教训 */
  lessons_learned: string;
  
  /** 预防措施 */
  prevention_measures?: string;
}

/**
 * 编辑总结请求接口
 */
export interface EditSummaryRequest {
  /** 根因分析 */
  root_cause?: string;
  
  /** 解决方案 */
  solution_approach?: string;
  
  /** 经验教训 */
  lessons_learned?: string;
  
  /** 预防措施 */
  prevention_measures?: string;
}

/**
 * 问题总结响应接口
 */
export interface IssueSummaryResponse {
  /** 是否成功 */
  success: boolean;
  
  /** 消息 */
  message: string;
  
  /** 总结数据 */
  data?: IssueSummary;
}

// ============================================================================
// 选项常量数组
// ============================================================================

/**
 * 状态选项常量数组
 * 用于下拉选择器和状态标签显示
 */
export const statusOptions: OptionItem[] = [
  { label: '新建', value: '新建', type: 'info' },
  { label: '处理中', value: '处理中', type: 'primary' },
  { label: '已解决', value: '已解决', type: 'success' },
  { label: '已关闭', value: '已关闭', type: '' },
];

/**
 * 优先级选项常量数组
 * 用于下拉选择器和优先级标签显示
 */
export const priorityOptions: OptionItem[] = [
  { label: '高', value: '高', type: 'danger' },
  { label: '中', value: '中', type: 'warning' },
  { label: '低', value: '低', type: 'success' },
];

/**
 * 分类选项常量数组
 * 可根据实际业务需求扩展
 */
export const categoryOptions: OptionItem[] = [
  { label: 'Bug', value: 'Bug', type: 'danger' },
  { label: '优化', value: '优化', type: 'primary' },
  { label: '需求', value: '需求', type: 'success' },
  { label: '其他', value: '其他', type: 'info' },
];

// ============================================================================
// 工具函数类型（可选）
// ============================================================================

/**
 * 获取状态标签类型
 * @param status 问题状态
 * @returns Element Plus Tag 类型
 */
export function getStatusTagType(status?: IssueStatus): TagType {
  if (!status) return '';
  const option = statusOptions.find(opt => opt.value === status);
  return option?.type || '';
}

/**
 * 获取优先级标签类型
 * @param priority 优先级
 * @returns Element Plus Tag 类型
 */
export function getPriorityTagType(priority?: IssuePriority): TagType {
  if (!priority) return '';
  const option = priorityOptions.find(opt => opt.value === priority);
  return option?.type || '';
}
